path_role.go

```diff
@@ -0,0 +1,346 @@
+	}
+
+	periodRaw, ok := data.GetOk("period")
+	if ok {
+		role.Period = time.Duration(periodRaw.(int)) * time.Second
+	} else if req.Operation == logical.CreateOperation {
+		role.Period = time.Duration(data.Get("period").(int)) * time.Second
+	}
+	if role.Period > b.System().MaxLeaseTTL() {
+		return logical.ErrorResponse(fmt.Sprintf("'period' of '%q' is greater than the backend's maximum lease TTL of '%q'", role.Period.String(), b.System().MaxLeaseTTL().String())), nil
+	}
+
+	if tokenNumUsesRaw, ok := data.GetOk("num_uses"); ok {
+		role.NumUses = tokenNumUsesRaw.(int)
+	} else if req.Operation == logical.CreateOperation {
+		role.NumUses = data.Get("num_uses").(int)
+	}
+	if role.NumUses < 0 {
+		return logical.ErrorResponse("num_uses cannot be negative"), nil
+	}
+
+	if tokenTTLRaw, ok := data.GetOk("ttl"); ok {
+		role.TTL = time.Duration(tokenTTLRaw.(int)) * time.Second
+	} else if req.Operation == logical.CreateOperation {
+		role.TTL = time.Duration(data.Get("ttl").(int)) * time.Second
+	}
+
+	if tokenMaxTTLRaw, ok := data.GetOk("max_ttl"); ok {
+		role.MaxTTL = time.Duration(tokenMaxTTLRaw.(int)) * time.Second
+	} else if req.Operation == logical.CreateOperation {
+		role.MaxTTL = time.Duration(data.Get("max_ttl").(int)) * time.Second
+	}
+
+	if boundAudiences, ok := data.GetOk("bound_audiences"); ok {
+		role.BoundAudiences = boundAudiences.([]string)
+	}
+
+	if boundSubject, ok := data.GetOk("bound_subject"); ok {
+		role.BoundSubject = boundSubject.(string)
+	}
+
+	if boundCIDRs, ok := data.GetOk("bound_cidrs"); ok {
+		parsedCIDRs, err := parseutil.ParseAddrs(boundCIDRs)
+		if err != nil {
+			return logical.ErrorResponse(err.Error()), nil
+		}
+		role.BoundCIDRs = parsedCIDRs
+	}
+
+	if userClaim, ok := data.GetOk("user_claim"); ok {
+		role.UserClaim = userClaim.(string)
+	}
+	if role.UserClaim == "" {
+		return logical.ErrorResponse("a user claim must be defined on the role"), nil
+	}
+
+	if groupsClaim, ok := data.GetOk("groups_claim"); ok {
+		role.GroupsClaim = groupsClaim.(string)
+	}
+
+	if len(role.BoundAudiences) == 0 &&
+		len(role.BoundCIDRs) == 0 &&
+		role.BoundSubject == "" {
+		return logical.ErrorResponse("must have at least one bound constraint when creating/updating a role"), nil
+	}
+
+	// Check that the TTL value provided is less than the MaxTTL.
+	// Sanitizing the TTL and MaxTTL is not required now and can be performed
+	// at credential issue time.
+	if role.MaxTTL > 0 && role.TTL > role.MaxTTL {
+		return logical.ErrorResponse("ttl should not be greater than max_ttl"), nil
+	}
+
+	var resp *logical.Response
+	if role.MaxTTL > b.System().MaxLeaseTTL() {
+		resp = &logical.Response{}
+		resp.AddWarning("max_ttl is greater than the system or backend mount's maximum TTL value; issued tokens' max TTL value will be truncated")
+	}
+
+	// Store the entry.
+	entry, err := logical.StorageEntryJSON("role/"+strings.ToLower(roleName), role)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
```

Comment:
Looking at `StorageEntryJSON`, I don't think this can ever be true.

Reply:
Fixed